src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -527,6 +528,13 @@ private ReadQuery getSliceCommands(QueryOptions options, ColumnFilter columnFilt
         if (keys.isEmpty())
             return ReadQuery.empty(table);
 
+        if (restrictions.keyIsInRelation())
+        {
+            Guardrails.partitionKeysInSelect.guard(keys.size(),
+                                                   String.format("partition keys in IN clause on table %s", table.name),
+                                                   null);
```

Comment:
Here we need to pass the `ClientState` for the query, so the guardrail is not applied to super user and internal queries. 

This would require to modify the signature of this method and also the signature of `SelectStatement#getQuery(QueryOptions, ColumnFilter, int, int)`. I just did that as part of the [PR](https://github.com/apache/cassandra/pull/1444/files) for [CASSANDRA-17187](https://issues.apache.org/jira/browse/CASSANDRA-17187), you can take the relevant changes on signatures from that patch.

Also, we would need a test to verify that the guardrail is skipped for those users. There are multiple tests named `testExcludedUsers` around that are used for equivalent checks for other guardrails.

Reply:
👍